### DISCUSSION TITLE
How do you order by a custom meta value?

### DISCUSSION THREAD

**Discussion by bmarshall511** [wp-graphql/wp-graphql#2389] [Q&A]
Done a ton of Googling and found very little on how to order by a custom meta key. I've seen people mention filters, but most are a few years old and don't seem to work anymore. If someone could shoot me some example code on how to sort posts by a custom meta key, think it'd be really useful to others.

**Comment by lstellway** [comment]
Hey @bmarshall511 
Very late now, but currently working to implement this on my site as well. 

You can look into the [`graphql_connection_query_args`](https://github.com/search?q=repo%3Awp-graphql%2Fwp-graphql%20graphql_connection_query_args&type=code) filter to manipulate the [`WP_Query::orderby`](https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters) parameter. This will likely need to be coupled with a filter for registering other enum values for your GraphQL connection `where` parameter. 

I have been integrated with the `wp-graphql` package for some time now and have an older filter in my project:

```php
add_filter('Products_orderby_enum_values', [$this, 'addProductOrderByValues']);
```

I don't recall where in the code that comes into play, but I am seeing a [`graphql_enum_values`](https://github.com/search?q=repo%3Awp-graphql%2Fwp-graphql+graphql_enum_values&type=code) filter that may work for this purpose. 
_Apologies, for not providing any readily available code, but hopefully this helps point others in the right direction._

**Comment by lstellway** [comment]
Okay - following up with my implementation:

First, I have a filter for adding an `orderby` enum value for the `Product` entity:
_(I am using [`wp-graphql-woocommerce`](https://github.com/wp-graphql/wp-graphql-woocommerce), but again, the [`graphql_enum_values`](https://github.com/search?q=repo%3Awp-graphql%2Fwp-graphql+graphql_enum_values&type=code) filter looks like it would be helpful here for other cases)_

```php
add_filter('Products_orderby_enum_values', function($values) {
    // In my example, I'm adding "stock status" as a valid value
    $values['STOCK_STATUS'] = [
        'value' => '_stock_status',
        'description' => __('Order by product\'s stock status', 'backroads'),
    ];

    return $values;
});
```

After the enum values are in place, we can implement the actual query filters using [`graphql_connection_query_args`](https://github.com/search?q=repo%3Awp-graphql%2Fwp-graphql%20graphql_connection_query_args&type=code):
_(related documentation: [WordPress > WP_Query > Order & Orderby Parameters](https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters))_

```php
$metaKeys = ['_stock_status'];

add_filter('graphql_connection_query_args', function($args) {
    $hasGraphqlOrderByClause = isset($args['graphql_args']['where']['orderby'])
        ? is_array($args['graphql_args']['where']['orderby'])
        : false;

    // Return early if we do not have an `orderby` clause
    if (!$hasGraphqlOrderByClause) {
        return $args;
    }

    // Ensure `orderby` is a keyed array
    if (!is_array($args['orderby'])) {
        $args['orderby'] = [
            $args['orderby'] => $args['order'] ?? 'ASC',
        ];
    }

    // [ 'field' => String, 'order' => 'ASC'|'DESC' ]
    $graphqlOrderbyClause = $args['graphql_args']['where']['orderby'];
    $defaultOrder = $args['order'] ?? 'ASC';

    foreach ($graphqlOrderbyClause as $orderBy) {
        $field = $orderBy['field'];
        $order = $orderBy['order'];
        $fieldClause = sprintf('%s_clause', $field);
        $isMetaField = in_array($field, $metaKeys);

        // Add order for standard fields
        if (!$isMetaField) {
            $args['orderby'][$field] = $order ?? $defaultOrder;
            continue;
        }

        // Handle order for meta fields
        if ($isMetaField) {
            // Initialize the 'meta_query' arg
            if (!isset($args['meta_query']) || !is_array($args['meta_query'])) {
                $args['meta_query'] = ['relation' => 'AND'];
            }

            // Add clase to the 'meta_query'
            $args['meta_query'][$fieldClause] = [
                'key' => $field,
                'compare' => 'EXISTS',
            ];

            // Add clause to 'orderby'
            $args['orderby'][$fieldClause] = $order ?? $defaultOrder;

            // Ensure a meta field order is not present without a corresponding query clause
            if (isset($args['orderby'][$field])) {
                unset($args['orderby'][$field]);
            }
        }
    }

    return $args;
});
```
